File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> list = new ArrayList<>();
        
        for (int i = 0; i < nums.length - 2; i++) {
            if (i == 0 || i > 0 && nums[i] != nums[i-1]) {
                int target = 0 - nums[i];
                int left = i + 1;
                int right = nums.length - 1;
                
                while (left < right) {
                    if (nums[left] + nums[right] == target) {
                        List<Integer> mini = new ArrayList<>();
                        mini.add(nums[i]);
                        mini.add(nums[left]);
                        mini.add(nums[right]);
                        list.add(mini);
                        
                        while (left < right && nums[left] == nums[left + 1]) left++;
                        while (left < right && nums[right] == nums[right - 1 ]) right--;
                        left++;
                        right--;
 
                    } else if (nums[left] + nums[right] > target) {
                        right--;
                    } else {
                        left++;
                    }
                }
            }
        }
        
        return list;
    }
}

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        
        for i, a in enumerate(nums):
            if i > 0 and a == nums[i - 1]:
                continue
            
            
            l = i + 1
            r = len(nums) - 1
            target = 0
            
            while l < r:
                current_sum = nums[l] + nums[r] + a
                if current_sum < target:
                    l += 1
                elif current_sum > target:
                    r -= 1
                else:
                    res.append([a, nums[l], nums[r]])
                    l += 1
                    while nums[l] == nums[l - 1] and l < r:
                        l += 1
        return res